### PR TITLE
[FLINK-17820][task][checkpointing] Respect fileSizeThreshold in FsCheckpointStateOutputStream.flush()

### PR DESCRIPTION
**Pull request by rkhachatryan**
## What is the purpose of the change

See FLINK-17820 for full problem description. In short, Flink ignores `state.backend.fs.memory-threshold` and always creates a file per subtask per checkpoint.

This PR makes `FsCheckpointStateOutputStream.flush()` to take `fileStateThreshold` into account.

## Verifying this change

This change added tests:
 - `FsCheckpointStreamFactoryTest.testFlushUnderThreshold`
 - `FsCheckpointStreamFactoryTest.testFlushAboveThreshold`
 - `ChannelStateCheckpointWriterTest.testSmallFilesNotWritten`

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): no
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: no
  - The serializers: no
  - The runtime per-record code paths (performance sensitive): no
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn/Mesos, ZooKeeper: no
  - The S3 file system connector: no

## Documentation

  - Does this pull request introduce a new feature? no
  - If yes, how is the feature documented? not applicable

### CODE REVIEW

File: flink-runtime/src/main/java/org/apache/flink/runtime/state/filesystem/FsCheckpointStreamFactory.java [hunk 1]
```diff
@@ -266,6 +264,16 @@ public void flush() throws IOException {
 			}
 		}
 
+		/**
+		 * Flush buffers to file if their size is above {@link #localStateThreshold}.
+		 */
+		@Override
+		public void flush() throws IOException {
+			if (pos > localStateThreshold) {
```
Comment (StephanEwen) [inline flink-runtime/src/main/java/org/apache/flink/runtime/state/filesystem/FsCheckpointStreamFactory.java:272]: Maybe extend this to `if (stream != null || pos > localStateThreshold)`?

File: flink-runtime/src/main/java/org/apache/flink/runtime/state/filesystem/FsCheckpointStreamFactory.java [hunk 2]
```diff
@@ -247,15 +247,13 @@ public long getPos() throws IOException {
 			return pos + (outStream == null ? 0 : outStream.getPos());
 		}
 
-		@Override
-		public void flush() throws IOException {
+		private void flushToFile() throws IOException {
```
Comment (StephanEwen) [inline flink-runtime/src/main/java/org/apache/flink/runtime/state/filesystem/FsCheckpointStreamFactory.java:250]: We can make this public, let the tests call it directly.

File: flink-runtime/src/test/java/org/apache/flink/runtime/state/filesystem/FsCheckpointStorageTest.java [hunk 3]
```diff
@@ -165,8 +165,7 @@ public void testDirectoriesForExclusiveAndSharedState() throws Exception {
 		CheckpointStateOutputStream exclusiveStream =
 				storageLocation.createCheckpointStateOutputStream(CheckpointedStateScope.EXCLUSIVE);
 
-		exclusiveStream.write(42);
-		exclusiveStream.flush();
```
Comment (StephanEwen) [inline flink-runtime/src/test/java/org/apache/flink/runtime/state/filesystem/FsCheckpointStorageTest.java:169]: If we type the `exclusiveStream` to `FsCheckpointStateOutputStream`, we could call `flushToFile()` directly.
Then we don't implicitly rely on the threshold logic here as well.

File: flink-runtime/src/main/java/org/apache/flink/runtime/state/filesystem/FsCheckpointStreamFactory.java (hunk 1 above)
Comment (rkhachatryan) [reply flink-runtime/src/main/java/org/apache/flink/runtime/state/filesystem/FsCheckpointStreamFactory.java:272]: Makes sense, expanded the condition.

File: flink-runtime/src/main/java/org/apache/flink/runtime/state/filesystem/FsCheckpointStreamFactory.java (hunk 2 above)
Comment (rkhachatryan) [reply flink-runtime/src/main/java/org/apache/flink/runtime/state/filesystem/FsCheckpointStreamFactory.java]: Done.

File: flink-runtime/src/test/java/org/apache/flink/runtime/state/filesystem/FsCheckpointStorageTest.java (hunk 3 above)
Comment (rkhachatryan) [reply flink-runtime/src/test/java/org/apache/flink/runtime/state/filesystem/FsCheckpointStorageTest.java:169]: Good point!

**Comment by rkhachatryan** [comment]
Thanks for the review @StephanEwen,
I've addressed the issues you raised.

**Comment by StephanEwen** [comment]
Thanks for the quick turnaround.

+1 to merge this from my side

File: flink-runtime/src/main/java/org/apache/flink/runtime/state/filesystem/FsCheckpointStreamFactory.java (hunk 1 above)
Comment (zhijiangW) [reply flink-runtime/src/main/java/org/apache/flink/runtime/state/filesystem/FsCheckpointStreamFactory.java:272]: I guess the added condition of `stream != null` is not addressed yet?

File: flink-runtime/src/test/java/org/apache/flink/runtime/checkpoint/channel/ChannelStateCheckpointWriterTest.java [hunk 4]
```diff
@@ -95,12 +118,12 @@ public void flush() throws IOException {
 
 		FlushRecorder dataStream = new FlushRecorder();
 		final ChannelStateCheckpointWriter writer = new ChannelStateCheckpointWriter(
-			1L,
-			new ChannelStateWriteResult(),
-			new ChannelStateSerializerImpl(),
-			NO_OP_RUNNABLE,
-			new MemoryCheckpointOutputStream(42),
-			dataStream
+				1L,
```
Comment (zhijiangW) [inline flink-runtime/src/test/java/org/apache/flink/runtime/checkpoint/channel/ChannelStateCheckpointWriterTest.java:121]: nit: unnecessary formatting changes, I guess it was done automatically by IDE.

File: flink-runtime/src/main/java/org/apache/flink/runtime/state/filesystem/FsCheckpointStreamFactory.java (hunk 1 above)
Comment (rkhachatryan) [reply flink-runtime/src/main/java/org/apache/flink/runtime/state/filesystem/FsCheckpointStreamFactory.java:272]: I must have reverted it while updating the tests. Fixed.

File: flink-runtime/src/test/java/org/apache/flink/runtime/checkpoint/channel/ChannelStateCheckpointWriterTest.java (hunk 4 above)
Comment (rkhachatryan) [reply flink-runtime/src/test/java/org/apache/flink/runtime/checkpoint/channel/ChannelStateCheckpointWriterTest.java]: Yes, accidental change, reverted it.
